test/index.js

```diff
@@ -183,6 +190,163 @@ test('`textDocument/didOpen`, `textDocument/didClose` (and diagnostics)', async
+  )
+  connection.sendNotification(DidOpenTextDocumentNotification.type, {
+    textDocument: {uri, languageId: 'markdown', version: 1, text: '# hi'}
+  })
+  await reopenDiagnosticsPromise
+  t.deepEqual(
+    configRequest,
+    {items: [{scopeUri: uri, section: 'remark'}]},
+    'should clear the cache if the file is opened'
+  )
+
+  configRequest = undefined
+  const changeConfigurationDiagnosticsPromise = createOnNotificationPromise(
+    connection,
+    PublishDiagnosticsNotification.type
+  )
+  requireConfig = true
+  connection.sendNotification(DidChangeConfigurationNotification.type, {
+    settings: {}
+  })
+  const changeConfigurationDiagnostics =
+    await changeConfigurationDiagnosticsPromise
+  t.deepEqual(
+    configRequest,
+    {items: [{scopeUri: uri, section: 'remark'}]},
+    'should clear the cache if the configuration changed'
+  )
+  t.deepEqual(
+    {uri, version: 1, diagnostics: []},
+    changeConfigurationDiagnostics,
+    'should not emit diagnostics if requireConfig is false'
+  )
+})
+
+test.only('global configuration `requireConfig`', async (t) => {
+  const connection = startLanguageServer(t, 'remark-with-warnings.js')
+  await connection.sendRequest(InitializeRequest.type, {
+    processId: null,
+    rootUri: null,
+    capabilities: {},
+    workspaceFolders: null
+  })
+
+  const uri = new URL('lsp.md', import.meta.url).href
+
+  const openDiagnosticsPromise = createOnNotificationPromise(
+    connection,
+    PublishDiagnosticsNotification.type
+  )
+  connection.sendNotification(DidOpenTextDocumentNotification.type, {
+    textDocument: {uri, languageId: 'markdown', version: 1, text: '# hi'}
+  })
+  const openDiagnostics = await openDiagnosticsPromise
+  t.notEqual(
+    openDiagnostics.diagnostics.length,
+    0,
+    'should emit diagnostics on `textDocument/didOpen`'
+  )
+
+  const changeConfigurationDiagnosticsPromise = createOnNotificationPromise(
+    connection,
+    PublishDiagnosticsNotification.type
+  )
+  connection.sendNotification(DidChangeConfigurationNotification.type, {
+    settings: {requireConfig: true}
+  })
+  const changeConfigurationDiagnostics =
+    await changeConfigurationDiagnosticsPromise
+  t.deepEqual(
+    {uri, version: 1, diagnostics: []},
+    changeConfigurationDiagnostics,
+    'should emit empty diagnostics if requireConfig is true without config'
+  )
+
+  const rcPath = new URL('.testremarkrc.json', import.meta.url)
+  t.teardown(() => fs.rm(rcPath, {force: true}))
+  await fs.writeFile(rcPath, '{}\n')
+  const watchedFileDiagnosticsPromise = createOnNotificationPromise(
+    connection,
+    PublishDiagnosticsNotification.type
+  )
+  connection.sendNotification(DidChangeWatchedFilesNotification.type, {
+    changes: []
+  })
+  const watchedFileDiagnostics = await watchedFileDiagnosticsPromise
+  t.isNot(
+    0,
+    watchedFileDiagnostics.diagnostics.length,
+    'should emit diagnostics if requireConfig is true with config'
+  )
```

Comment:
I don’t understand why this doesn’t work. The idea is that changing the rc configuration file retriggers validation, which is does, and then finds the configuration file, but it doesn’t.

Reply:
To clarify, with “but it doesn’t” you mean “which it doesn’t”, right? So the trigger works, then it can’t find the file again?

Reply:
This test is *suuuuper* long btw, really hard to get what’s happening (or supposed to happen) 😅

Reply:
Yes, I agree it’s long. The other test is even longer. I tried grouping it using blank lines, but it’s still quite messy.

Basically 3 things happen:

0. The server is started and initialized without workspace support.
1. The file is opened. As a result file diagnostics are emitted (`diagnostics.length !== 0`)
2. The language server changes its option `requireConfig` to `true`. `.testremarkrc.json` is not present, so empty diagnostics are emitted.
3. The file `/home/remco/Projects/unified-language-server/test/.testremarkrc.json` file is created and the server is notified of this. This should result in diagnostics again, but the diagnostics are empty.
   These are the values passed into `unified-engine` (inspected using a debugger, then tweaked manually for representation here)
   ```js
   {
     alwaysStringify: false,
     cwd: '/home/remco/Projects/unified-language-server',
     files: [
       // VFile simplified for representation
       {
         messages: [],
         path: '/home/remco/Projects/unified-language-server/test/lsp.md',
         cwd: '/home/remco/Projects/unified-language-server',
         value: '# hi'
       }
     ],
     ignoreUnconfigured: true,
     plugins: [ './test/lots-of-warnings.js' ],
     quiet: false,
     rcName: 'testremark',
     silentlyIgnore: true,
     streamError: new PassThrough(),
     streamOut: new PassThrough()
   }
   ```
   The language server used is `remark-with-warnings.js`:
   ```js
   import {createUnifiedLanguageServer} from '../index.js'

   createUnifiedLanguageServer({
     configurationSection: 'remark',
     processorName: 'remark',
     processorSpecifier: 'remark',
     rcName: 'testremark',
     // This is resolved from the directory containing package.json
     plugins: ['./test/lots-of-warnings.js']
   })
   ```
   I really don’t understand why this yields empty diagnostics